src/txrequest.cpp

```diff
@@ -0,0 +1,715 @@
+            // Can have at most 1 CANDIDATE_BEST/REQUESTED peer
+            assert(c.m_candidate_best + c.m_requested <= 1);
+            // If there are any CANDIDATE_READY entries, there must be exactly one CANDIDATE_BEST or REQUESTED
+            // entry.
+            if (c.m_candidate_ready > 0) {
+                assert(c.m_candidate_best + c.m_requested == 1);
+            }
+            // If there is both a CANDIDATE_READY and a CANDIDATE_BEST entry, the CANDIDATE_BEST one must be at
+            // least as good as the best CANDIDATE_READY.
+            if (c.m_candidate_ready && c.m_candidate_best) {
+                assert(c.m_priority_candidate_best <= c.m_priority_best_candidate_ready);
+            }
+            // Detect duplicate (peer, txid) entries
+            std::sort(c.m_peers.begin(), c.m_peers.end());
+            assert(std::adjacent_find(c.m_peers.begin(), c.m_peers.end()) == c.m_peers.end());
+            // Verify all per_txhash flags.
+            uint8_t expected_per_txhash = 0;
+            if (c.m_any_preferred_first || c.m_requested) {
+                expected_per_txhash |= TXHASHINFO_NO_MORE_PREFERRED_FIRST;
+            }
+            if (c.m_any_nonpreferred_first || c.m_requested) {
+                expected_per_txhash |= TXHASHINFO_NO_MORE_NONPREFERRED_FIRST;
+            }
+            // All expected flags must be present, but there can be more. If a node went from REQUESTED to
+            // COMPLETED, or was deleted, our expected_per_txhash may miss the relevant bits.
+            assert((expected_per_txhash & ~c.m_or_all_per_txhash) == 0);
+            // No entry can have flags that are a superset of the actual ones (they're always ORed into the actual
+            // one).
+            auto it_last = std::prev(m_index.get<ByTxHash>().lower_bound(
+                EntryTxHash{entry.first, State::TOO_LARGE, 0}));
+            assert(it_last->m_txhash == entry.first);
+            assert(c.m_or_all_per_txhash == it_last->m_per_txhash);
+        }
+    }
+
+    void PostGetRequestableSanityCheck(std::chrono::microseconds now) const
+    {
+        for (const auto& entry : m_index) {
+            if (entry.IsWaiting()) {
+                // REQUESTED and CANDIDATE_DELAYED must have a time in the future (they should have been converted
+                // to COMPLETED/CANDIDATE_READY respectively).
+                assert(entry.m_time > now);
+            } else if (entry.IsSelectable()) {
+                // CANDIDATE_READY and CANDIDATE_BEST cannot have a time in the future (they should have remained
+                // CANDIDATE_DELAYED, or should have been converted back to it if time went backwards).
+                assert(entry.m_time <= now);
+            }
+        }
+    }
+};
+
+TxRequestTracker::TxRequestTracker(bool deterministic)
+{
+    m_impl = MakeUnique<TxRequestTracker::Impl>(deterministic);
+}
+
+TxRequestTracker::~TxRequestTracker() {}
```

Comment:
Consider using the `default` syntax to indicate that you only included this in the cpp file so the `m_impl` unique ptr can be destructed:

```suggestion
TxRequestTracker::~TxRequestTracker() = default;
```

Reply:
Of course! Done.